File: mjtp/mjtp.go
package mjtp

import (
	"encoding/json"
	"errors"
	"strings"
)

type Message struct {
	Resource string
	Version  string
	Body     map[string]interface{}
}

func Make(resource string, body map[string]interface{}) *Message {
	return &Message{
		Resource: resource,
		Version:  "MJTP/1.0",
		Body:     body,
	}
}

func (m *Message) String() (string, error) {
	body := "{}"

	if m.Body != nil {
		j, err := json.Marshal(m.Body)

		if err != nil {
			return "", err
		}
		body = string(j)
	}

	return m.Resource + " " + m.Version + " " + body + "\r\n\r\n", nil
}

func Parse(data string) (*Message, error) {
	message := &Message{}
	message.Body = make(map[string]interface{})

	buffer := data
	bufferIdx := 0

	// Find the first space
	bufferIdx = strings.Index(buffer, " ")
	message.Resource = buffer[:bufferIdx]

	// Find the second space
	buffer = buffer[bufferIdx+1:]
	bufferIdx = strings.Index(buffer, " ")
	message.Version = buffer[:bufferIdx]

	// Find the first \r\n\r\n
	buffer = buffer[bufferIdx+1:]
	bufferIdx = strings.Index(buffer, "\r\n\r\n")

	if err := json.Unmarshal([]byte(buffer[:bufferIdx]), &message.Body); err != nil {

		return nil, err
	}
	if message.Version != "MJTP/1.0" {
		return nil, errors.New("invalid version")
	}

	return message, nil
}

File: www/src/mjtp.ts
export class MJTP {
	private resource: string;
	private version: string;
	private body: { [key: string]: any };

	constructor(resource: string, body: { [key: string]: any }) {
		this.resource = resource;
		this.version = "MJTP/1.0";
		this.body = body;
	}

	toString(): string {
		return `${this.resource} ${this.version} ${JSON.stringify(this.body)}\r\n\r\n`;
	}

	static parse(data: string): MJTP {
		// validate
		const parts = data.split(" ");

		if (parts.length < 3) {
			throw new Error("invalid message format");
		} else if (parts[1] !== "MJTP/1.0") {
			throw new Error("invalid version");
		} else if (parts[0].length === 0) {
			throw new Error("invalid resource");
		}

		let buffer: string = data
		let bufferPosition: number

		const mjtp = new MJTP('', {})

		bufferPosition = buffer.indexOf(' ')
		mjtp.resource = buffer.slice(0, bufferPosition)

		buffer = buffer.slice(bufferPosition + 1)
		bufferPosition = buffer.indexOf(' ')
		mjtp.version = buffer.slice(0, bufferPosition)

		buffer = buffer.slice(bufferPosition + 1)
		mjtp.body = JSON.parse(buffer.split('\r\n\r\n')[0])
		
		return mjtp
	}
}
